### ISSUE TITLE
Improved schema and catalog support from SQL

### ISSUE THREAD

**Issue by matthewmturner** [apache/arrow-datafusion#1877]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

Right now there is limited schema and catalog support in sql.  You have to be using language bindings and register the catalog / schema to the `ExecutionContext` before being able to select.  I would like to be able to create schemas and catalogs from sql, without language bindings, so that its easier to manage ddl and do data analysis just using datafusion-cli.
 
```
DataFusion CLI v7.0.0
❯ CREATE TABLE sch.tbl AS VALUES (1,2,3);
0 rows in set. Query took 0.032 seconds.
❯ show tables;
+---------------+--------------------+------------+------------+
| table_catalog | table_schema       | table_name | table_type |
+---------------+--------------------+------------+------------+
| datafusion    | public             | sch.tbl    | BASE TABLE |
| datafusion    | information_schema | tables     | VIEW       |
| datafusion    | information_schema | columns    | VIEW       |
+---------------+--------------------+------------+------------+
3 rows in set. Query took 0.015 seconds.
❯ CREATE TABLE cat.sch.tbl AS VALUES (1,2,3);
0 rows in set. Query took 0.003 seconds.
❯ show tables;
+---------------+--------------------+-------------+------------+
| table_catalog | table_schema       | table_name  | table_type |
+---------------+--------------------+-------------+------------+
| datafusion    | public             | cat.sch.tbl | BASE TABLE |
| datafusion    | public             | sch.tbl     | BASE TABLE |
| datafusion    | information_schema | tables      | VIEW       |
| datafusion    | information_schema | columns     | VIEW       |
+---------------+--------------------+-------------+------------+
4 rows in set. Query took 0.006 seconds.
```

```
❯ CREATE SCHEMA abc;
NotImplemented("Unsupported SQL statement: CreateSchema { schema_name: ObjectName([Ident { value: \"abc\", quote_style: None }]), if_not_exists: false }")
```

```
CREATE CATALOG xyz;  🤔 Invalid statement: sql parser error: Expected an object type after CREATE, found: CATALOG
```

**Describe the solution you'd like**
Support the following sql statements:
- `CREATE CATALOG abc`
- `CREATE SCHEMA xyz`
- Parse catalog and schema from `CREATE TABLE abc.xyz.tbl` instead of creating table `

- [x] Create Schema support
- [x] Create Catalog / Database support
- [ ] `/c` command on datafusion cli to set active db / catalog

**Describe alternatives you've considered**

**Additional context**
Postgres reference here
https://www.postgresql.org/docs/9.3/sql-createschema.html

**Comment by matthewmturner** [comment]
@houqp @Jimexist FYI - so you have an idea where im trying to go...im trying to make datafusion-cli a really quick and powerful tool for doing ad hoc data analysis / query testing - and i think improved catalog & schema support will help with that.  Through a combination of this, #1872, and im hoping to also eventually add an s3 feature (but i have plenty to work on before that) i hope to achieve that. I just wanted to give you both some additional color about my longer term objectives.

Let me know if any concerns.

**Comment by bobtins** [comment]
Hi, noticed this issue and just wanted to put my two cents in.
<soapbox>
The tl;dr is that most of the time you don't need these added levels; at most maybe add schema. 
I've worked with a data analysis tool that was cross-database in nature, and I noticed that catalogs and schemas mean different things for each DB impl, and most of the time I never had to deal with "catalog". If for PostgreSQL it's the same as "database", I'm not sure why you need "create catalog" if you have "create database".
The sample output you are showing looks like it's from "information_schema" which is just a generic metadata standard that tries to be one size fits all; as such, it has things that are only used in some DBs. That said, it's kind of counterintuitive that a "." in a table name is a legal character, as you demonstrate.
Personally, I can get by without either schema or catalog, but YMMV...
</soapbox>

**Comment by matthewmturner** [comment]
Fortunately, we already have the ability to parse table refs into catalog, schema, and table with `TableReference::resolve(default_catalog, default_schema) -> ResolvedTableReference`.  I havent seen any issues with this yet.

Definitely agree on maintaining current active catalog / schema.  From a pure sql / datafusion-cli perspective i was thinking of using the `\c` command like psql for connecting to a database / catalog.  Then for schema maybe support for the `SET` command with options like `search_path`.  Personally, im ok with those being follow on steps as im fine with using fully qualified names in queries for now.

i believe postgres supports putting databases names, schemas, and tables in double quotes so i would think we could support something similar for resolving names with ".".

In my current PR #1959 i thought i was actually quite close with adding `CREATE SCHEMA` support however im currently having an issue with getting the schema to register to the catalog.  Still need to figure that out.

**Comment by bobtins** [comment]
In this example, it seemed like parsing the table refs wasn't working:
```
❯ CREATE TABLE sch.tbl AS VALUES (1,2,3);
0 rows in set. Query took 0.032 seconds.
❯ show tables;
+---------------+--------------------+------------+------------+
| table_catalog | table_schema       | table_name | table_type |
+---------------+--------------------+------------+------------+
| datafusion    | public             | sch.tbl    | BASE TABLE |
```
Then I saw a change in datafusion/src/catalog/mod.rs to split the table reference; assuming that fixes it.

You said 
> Personally, im ok with those being follow on steps as im fine with using fully qualified names in queries for now.

I'm assuming we will still be able to use simple table names, right?
Take what I say with a grain of salt--I'm a little obsessive about keeping things simple, but I don't know the whole history. It sounds like you are making this feature more robust, and that's always good :)

**Comment by matthewmturner** [comment]
ah yeah, i was referring to in my PR (#1959) which was leveraging on what looked like an incomplete feature for properly resolving catalog / schema / table.  You can still query for single table names in the currently active catalog / schema - which right now is only 'datafusion' and 'public' - but i plan on adding a feature (such as psql \c to connect to database) to help with that.

I made significant progress on the above referenced PR, just working on the ballista serialization now.  Would be good to get another set of eyes on it if youre interested.

```
❯ CREATE SCHEMA abc;
0 rows in set. Query took 0.001 seconds.
❯ CREATE TABLE abc.y AS VALUES (1,2,3);
0 rows in set. Query took 0.002 seconds.
❯ SHOW TABLES;
+---------------+--------------------+----------------+------------+
| table_catalog | table_schema       | table_name     | table_type |
+---------------+--------------------+----------------+------------+
| datafusion    | public             | alltypes_plain | BASE TABLE |
| datafusion    | public             | x              | BASE TABLE |
| datafusion    | abc                | y              | BASE TABLE |
| datafusion    | information_schema | tables         | VIEW       |
| datafusion    | information_schema | columns        | VIEW       |
+---------------+--------------------+----------------+------------+
5 rows in set. Query took 0.016 seconds.
❯ SELECT * FROM x;
+---------+---------+
| column1 | column2 |
+---------+---------+
| 1       | 2       |
+---------+---------+
1 row in set. Query took 0.004 seconds.
❯ SELECT * FROM y;
Plan("Table or CTE with name 'y' not found")
❯ SELECT * FROM abc.y;
+---------+---------+---------+
| column1 | column2 | column3 |
+---------+---------+---------+
| 1       | 2       | 3       |
+---------+---------+---------+
1 row in set. Query took 0.005 seconds.
```

**Comment by alamb** [comment]
I think this is now complete. Let's file additional tickets for follow on work

```sql
DataFusion CLI v26.0.0
❯ create schema foo;
0 rows in set. Query took 0.019 seconds.
❯ create database bar;
0 rows in set. Query took 0.000 seconds.
```
